### DISCUSSION TITLE
Hard syncing products, collections, and variants

### DISCUSSION THREAD

**Discussion by JonKaric** [statamic-rad-pack/shopify#206] [Feature Requests]
It would be great to have an option to create a hard sync - whereby products, variants, tags etc are deleted if they don't exist in the incoming import. 

Currently, if you delete a product from Shopify and import products, the old products are still visible in Statamic.

**Comment by ryanmitchell** [comment]
Alternatively we could add a console task to clear all products, tags, vendor and collection data? Then you could run the import again?

**Reply by JonKaric** [reply]
That's good - I think I'd prefer it as a config setting though. Flick it on to have that data cleared automatically when you import. What do you think?

**Reply by ryanmitchell** [reply]
The config only makes sense to me if you would want it to clear it every time... I cant think of a use case for that scenario.

**Reply by JonKaric** [reply]
I'd want to clear it every time, especially for bigger catalogs. The expected behaviour for the layperson would be that their products collection exactly matches their Shopify product catalog, but that isn't what happens behind the scenes. 

If we can't set it on a config level, then it would be better to combine the deletion task and import task into one button and give a short description of what both do:

**Import**
_Imports new products and updates existing ones_

**Import & Sync**
_Imports new products, updates existing ones, and deletes products that are no longer in your Shopify catalog._

**Reply by ryanmitchell** [reply]
I don't agree thats the expected behaviour as it would overwrite any data you've created locally in Statamic. It definitely wouldn't be something we would want to happen on our installs.

But I understand your desire for it and why, when I get some time I'll work on it using the config approach and update you here. Feel free to open a PR if you have time.

**Reply by JonKaric** [reply]
> as it would overwrite any data you've created locally in Statamic

(Wouldn't want this to happen, just for it to clear products that no longer exist).

I'd basically want the same functionality as the variants as they're deleted when they no longer exist. Perhaps add the product to a cache and call upon it when the whole import job has been done, and delete the statamic products that don't exist in that cache. Something like that.

**Reply by ryanmitchell** [reply]
I think this is getting over complicated. Would it not just be better anfter updating and deleting to check the linked taxonomies (collections, vendor etc) and for any that have items without entries linked we delete them. That gives you what you are looking for from what I understand.

**Reply by JonKaric** [reply]
What if you don't have any taxonomies attached to the product?

**Reply by ryanmitchell** [reply]
Then theres nothing to clear up? 

As it stands, a product will be deleted through the appropriate web hook so we should check for any taxonomies that are empty. When a product is edited or the bulk import is run, we should also check for any taxonomies that are empty. Asides from that I dont think theres anything to do.

**Comment by JonKaric** [comment]
Closing as we've discussed on Discord it's a big change for products, might be able to PR at some point 😄
